proposals/0000-no-sugared-incompleteness.rst

```diff
@@ -0,0 +1,211 @@
+  The idea isn't that ``PatternMatchFail`` is actually good, but just to be consistent with the rest of the language and avoid using a exhaustiveness heuristic weaker than the regular exhaustiveness checker.
+  The combination of ``-XNoFallibleDo`` and ``-XNoIncomplete`` would make those incomplete patterns errors like all the others under ``-XNoIncomplete``.
+  This keeps the "knobs" orthogonal, and also keeps the language consistent whether ``Incomplete`` is disabled or not.
+
+Costs and Drawbacks
+-------------------
+
+- While Haskell 2010 works well with a simple completeness checker that wouldn't be too hard to specify, extensions like ``GADTs`` that could be standardized in the future immediately ramp up the desire for a far more sophisticated pattern match completeness checker.
+  Standardizing ``Incomplete`` could make it harder standardize ``GADTs`` and those other extensions later.
+
+  See the future work section for how this is mitigated.
+
+- This proposal effectively promotes some GHC warnings to a language extension.
+  While GHC strives for comparability, ensuring programs with ``-Werror=...``, once accepted, continue to be accepted in future releases was lower down on the prioritization of aspects of compatibility GHC sought to ensure.
+  This makes the completeness checker improvements changes more risky, as they now entail somewhat stronger comparability guarantees.
+
+  See the future work section for how this is mitigated.
+
+- In the short term, we have far too many knobs to control the same thing with each new ``-fdefer-*`` flag corresponding to an existing warning.
+  But, my hope is that in future versions of the language ``Incomplete`` can be deprecated and then removed, reducing the number of knobs back to something sane.
+
+- Somebody is going to think this has something to do with Kurt Gödel unless we choose a different name.
+
+Possible Future Work
+--------------------
+
+The drawbacks are serious enough that I feel compelled to sketch some future plans that would mitigate them.
+
+The chief drawback with pattern matching today as it relates to completeness checking is that the Haskell implementation cannot tell whether a user skipped a pattern because they think it's impossible or because they forgot.
+If we had something like Agda's "absurd patterns", however, the user could use those to use those cover the impossible the cases, making clear that anything else the user really did forget.
+Firstly, this should help make code more self-documenting and error/warning messages better.
+But especially relevant to the problem at hand, this allows simpler Haskell implementations that aren't so sophisticated that they can derive proofs of pattern impossibility very well on their own, but can verify user-written arguments.
+This allows extension conjunctions like ``NoIncomplete`` and ``GADTs`` to be specified in ways that are less onerous on the Haskell implementation.
```

Comment:
While Agda-style absurd patterns would be nice, I don't think they actually help much here, because you have to specify when they can be used. An absurd pattern on `x` is essentially equivalent to requiring that `case x of {}` covers all cases, so checking it doesn't make the implementation simpler.

In Agda, one has to use an absurd pattern when an impossible pattern is nested somewhere inside a larger pattern, but in Haskell such cases are normally trivially accessible because of laziness.

Reply:
I guess I should rephrase this to make it clear I don't mean exactly like Agda absurd patterns, but something along those lines.

> In Agda, one has to use an absurd pattern when an impossible pattern is nested somewhere inside a larger pattern, but in Haskell such cases are normally trivially accessible because of laziness.

Yes when it's just a nested `Void`-like thing, Haskell unintentionally makes one "prove it" which is great. But in the case GADTs type constraint stuff, Haskell is magical again. We'll soon (:crossed_fingers:) have type applications in patterns, I think I would want to put something in there (think nested absurd patterns for *irrelevant* fields. which aren't lazy) to get closer to an explicit proof for deriving the type inequalities.